src/repr/type_binary.ml

```diff
@@ -244,8 +241,62 @@ module Decode = struct
         unstage decoders.(i) buf ofs)
 end
 
+module Pre_hash = struct
+  type 'a pre_hash = 'a encode_bin
+
+  let rec t : type a. a t -> a pre_hash = function
+    | Self s -> self s
+    | Custom c -> c.pre_hash
+    | Map m -> map m
+    | Boxed b -> t b
+    | Attributes { attr_type; _ } -> t attr_type
+    | List l -> Encode.list (t l.v) l.len
+    | Array a -> Encode.array (t a.v) a.len
+    | Tuple t -> tuple t
+    | Option x -> Encode.option (t x)
+    | Record r -> record r
+    | Variant v -> variant v
+    | Var v -> raise (Unbound_type_variable v)
+    | Prim _ as t -> Encode.unboxed t
```

Comment:
Couldn't that unboxing be source of preimage attacks? 

Both `("a", "b")` and `("ab", "")` would prehash the same way.

Reply:
Agreed, this is a problem. (Even aside from the security concerns, it's an changes the way hashes are computed in a non-backwards-compatible way.)

We still need outermost `Prim _` cells to be encoded with `unboxed:true`, for backwards compatibility. I'll change this back & add some tests for the preimage attack problem.

Reply:
Now done in a new commit.